Automattic/woocommerce-payments
# Add/1982 manage payment methods upe

Fixes #1982

#### Changes proposed in this Pull Request









This PR adds the option to save a payment method on My Account > Payment Methods > Add Payment Method using Universal Payment Method (UPE).

Payment methods are saved using Setup Intents. 
The workflow:
1) When customer clicks on Add Payment Method and selects WooCommerce Payments, a setup intent is created, without confirmation. The setup Intent ID is used to create the UPE element.
2) Customer enters details and clicks on Add Payment Method. The setup intent is confirmed and there is a redirection to the screen Manage Payment Methods.
3) On Manage Payment Methods, we clear the cache for the customer's payment methods and retrieve the new list from Stripe. We display a notice that the payment was successfully created.

Not on this PR:
- Styling UPE form for Add Payment Method screen.
- Managing SEPA and SOFORT payment methods.

#### Testing instructions





* With UPE Enabled (Enter your WC Options table and add the feature flag `_wcpay_feature_upe` with a value of 1) go to My Account > Manage Payments > Add Payment Method
* Select WooCommerce Payments, verify UPE iFrame is displayed

<img width="709" alt="UPEMyAccount" src="https://user-images.githubusercontent.com/8002881/121905584-ed92c100-ccef-11eb-86c6-b73817244630.png">
* Enter incomplete information, click on Add Payment Method, check Errors are displayed
<img width="692" alt="UPEErrors" src="https://user-images.githubusercontent.com/8002881/121905783-1d41c900-ccf0-11eb-8173-f4fd7d5e3fbc.png">
* Enter all the fields, click on Add Payment Method. Check that you are redirected to Manage Payments, there is a notice indicating the Payment Method was added and that you see it on the list.

<img width="915" alt="GreatSuccess" src="https://user-images.githubusercontent.com/8002881/121906102-65f98200-ccf0-11eb-96aa-03a81d6cd24b.png">
* Check that you can checkout using the newly saved payment method.
* Go back to Manage Payment Methods, check that you can set the payment method as default and delete it.
-------------------

- [x] Tested on mobile (or does not apply) Not comprehensively.
